File: alljoyn_unity/unit_test/ObjectSecurityTest.cs
//-----------------------------------------------------------------------
// <copyright file="ObjectSecurityTest.cs" company="AllSeen Alliance.">
// Copyright (c) 2013, AllSeen Alliance. All rights reserved.
//
//    Permission to use, copy, modify, and/or distribute this software for any
//    purpose with or without fee is hereby granted, provided that the above
//    copyright notice and this permission notice appear in all copies.
//
//    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
//    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
//    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
//    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
//    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
//    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
//    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
// </copyright>
//-----------------------------------------------------------------------

using System;
using System.Threading;
using System.Collections.Generic;
using System.Text;
using AllJoynUnity;
using Xunit;

namespace AllJoynUnityTest
{
	public class ObjectSecurityTest : IDisposable
	{
		const string INTERFACE_NAME = "org.alljoyn.test.cs.objsecurity";
		const string OBJECT_PATH = "/org/alljoyn/test/objsecurity";
		const string WELLKNOWN_NAME = "org.alljoyn.test.cs.objsecurity";

		private AllJoyn.BusAttachment serviceBus;
		private AllJoyn.BusAttachment clientBus;
		private TestBusObject busObject;


		public struct Authflags
		{
			public bool requestCreds_service;
			public bool authComplete_serivce;
			public bool verifyCreds_service;
			public bool securityViolation_service;

			public bool requestCreds_client;
			public bool authComplete_client;
			public bool verifyCreds_client;
			public bool securityViolation_client;
		}

		public Authflags authflags;
		public bool nameOwnerChangedFlag;

		//thread to notify when callbacks have occured.
		private AutoResetEvent notifyEvent = new AutoResetEvent(false);
		private void Wait(TimeSpan timeout)
		{
			notifyEvent.WaitOne(timeout);
			notifyEvent.Reset();
		}

		private void Notify()
		{
			notifyEvent.Set();
		}

		private bool disposed = false;

		public ObjectSecurityTest()
		{
			serviceBus = new AllJoyn.BusAttachment("AuthListenerTestService", false);

			Assert.Equal(AllJoyn.QStatus.OK, serviceBus.Start());
			Assert.Equal(AllJoyn.QStatus.OK, serviceBus.Connect(AllJoynTestCommon.GetConnectSpec()));

			clientBus = new AllJoyn.BusAttachment("AuthListenerTestClient", false);

			Assert.Equal(AllJoyn.QStatus.OK, clientBus.Start());
			Assert.Equal(AllJoyn.QStatus.OK, clientBus.Connect(AllJoynTestCommon.GetConnectSpec()));
		}

		~ObjectSecurityTest()
		{
			Dispose(false);
		}

		protected virtual void Dispose(bool disposing)
		{
			if (!this.disposed)
			{
				// additional clean up before disposing
				serviceBus.Stop();
				serviceBus.Join();

				clientBus.Stop();
				clientBus.Join();

				if (disposing)
				{
					serviceBus.Dispose();
					clientBus.Dispose();
				}
				disposed = true;
			}
		}

		public void Dispose()
		{
			Dispose(true);
			GC.SuppressFinalize(this);
		}

		void ResetAuthFlags()
		{
			authflags.requestCreds_service = false;
			authflags.authComplete_serivce = false;
			authflags.verifyCreds_service = false;
			authflags.securityViolation_service = false;

			authflags.requestCreds_client = false;
			authflags.authComplete_client = false;
			authflags.verifyCreds_client = false;
			authflags.securityViolation_client = false;
		}

		class TestBusListener : AllJoyn.BusListener
		{
			ObjectSecurityTest _objectSecurityTest;

			public TestBusListener(ObjectSecurityTest objectSecurityTest)
			{
				this._objectSecurityTest = objectSecurityTest;
			}

			protected override void NameOwnerChanged(string busName, string previousOwner, string newOwner)
			{
				_objectSecurityTest.nameOwnerChangedFlag = true;
				_objectSecurityTest.Notify();
			}
		}

		class TestBusObject : AllJoyn.BusObject
		{
			public TestBusObject(string path) : base(path) { }

			public void Ping(AllJoyn.InterfaceDescription.Member member, AllJoyn.Message message)
			{
				string outStr = (string)message[0];
				AllJoyn.MsgArg outArgs = new AllJoyn.MsgArg();
				outArgs = outStr;

				AllJoyn.QStatus status = MethodReply(message, outArgs);
				Assert.Equal(AllJoyn.QStatus.OK, status);
			}
		}


		public AllJoyn.QStatus SetUpAuthService()
		{
			AllJoyn.QStatus status = AllJoyn.QStatus.FAIL;

			TestBusListener busListener = new TestBusListener(this);
			serviceBus.RegisterBusListener(busListener);

			busObject = new TestBusObject(OBJECT_PATH);
			AllJoyn.InterfaceDescription ifaceDescritpion = serviceBus.GetInterface(INTERFACE_NAME);
			status = busObject.AddInterface(ifaceDescritpion);
			if (!status)
			{
				return status;
			}

			AllJoyn.InterfaceDescription.Member ping_member = ifaceDescritpion.GetMethod("ping");

			status = busObject.AddMethodHandler(ping_member, busObject.Ping);
			if (!status)
			{
				return status;
			}

			status = serviceBus.RegisterBusObject(busObject, true);
			if (!status)
			{
				return status;
			}

			Assert.True(busObject.IsSecure);

			status = serviceBus.RequestName(WELLKNOWN_NAME,
				AllJoyn.DBus.NameFlags.ReplaceExisting |
				AllJoyn.DBus.NameFlags.DoNotQueue |
				AllJoyn.DBus.NameFlags.AllowReplacement);
			if (!status)
			{
				return status;
			}
			Wait(TimeSpan.FromSeconds(5));
			return status;
		}

		class SrpKeyx_service_authlistener : AllJoyn.AuthListener
		{
			private ObjectSecurityTest _objectSecurityTest;
			public SrpKeyx_service_authlistener(ObjectSecurityTest objectSecurityTest)
			{
				_objectSecurityTest = objectSecurityTest;
			}
			protected override bool RequestCredentials(string authMechanism, string peerName,
				ushort authCount, string userName, AllJoyn.Credentials.CredentialFlags credMask,
				AllJoyn.Credentials credentials)
			{
				if ((credMask & AllJoyn.Credentials.CredentialFlags.Password) == AllJoyn.Credentials.CredentialFlags.Password)
				{
					credentials.Password = "ABCDEFGH";
				}
				_objectSecurityTest.authflags.requestCreds_service = true;
				return true;
			}

			protected override bool VerifyCredentials(string authMechanism, string peerName, AllJoyn.Credentials credentials)
			{
				_objectSecurityTest.authflags.verifyCreds_service = true;
				return true;
			}

			protected override void SecurityViolation(AllJoyn.QStatus status, AllJoyn.Message msg)
			{
				_objectSecurityTest.authflags.securityViolation_service = true;
			}

			protected override void AuthenticationComplete(string authMechanism, string peerName, bool success)
			{
				_objectSecurityTest.authflags.authComplete_serivce = true;
			}
		}

		class SrpKeyx_client_authlistener : AllJoyn.AuthListener
		{
			private ObjectSecurityTest _objectSecurityTest;
			public SrpKeyx_client_authlistener(ObjectSecurityTest objectSecurityTest)
			{
				_objectSecurityTest = objectSecurityTest;
			}

			protected override bool RequestCredentials(string authMechanism, string peerName,
				ushort authCount, string userName, AllJoyn.Credentials.CredentialFlags credMask,
				AllJoyn.Credentials credentials)
			{
				if ((credMask & AllJoyn.Credentials.CredentialFlags.Password) == AllJoyn.Credentials.CredentialFlags.Password)
				{
					credentials.Password = "ABCDEFGH";
				}
				_objectSecurityTest.authflags.requestCreds_client = true;
				return true;
			}

			protected override bool VerifyCredentials(string authMechanism, string peerName, AllJoyn.Credentials credentials)
			{
				_objectSecurityTest.authflags.verifyCreds_client = true;
				return true;
			}

			protected override void SecurityViolation(AllJoyn.QStatus status, AllJoyn.Message msg)
			{
				_objectSecurityTest.authflags.securityViolation_client = true;
			}

			protected override void AuthenticationComplete(string authMechanism, string peerName, bool success)
			{
				_objectSecurityTest.authflags.authComplete_client = true;
			}
		}

		[Fact, Trait("objsecurity", "test")]
		public void insecure_interface_secureobject()
		{
			AllJoyn.InterfaceDescription service_intf = null;
			Assert.Equal(AllJoyn.QStatus.OK, serviceBus.CreateInterface(INTERFACE_NAME, out service_intf));
			Assert.Equal(AllJoyn.QStatus.OK, service_intf.AddMethod("ping", "s", "s", "in,out"));
			service_intf.Activate();

			ResetAuthFlags();
			clientBus.ClearKeyStore();

			SrpKeyx_service_authlistener serviceAuthlistener = new SrpKeyx_service_authlistener(this);
			Assert.Equal(AllJoyn.QStatus.OK, serviceBus.EnablePeerSecurity("ALLJOYN_SRP_KEYX", serviceAuthlistener, null, false));

			serviceBus.ClearKeyStore();

			Assert.Equal(AllJoyn.QStatus.OK, SetUpAuthService());

			SrpKeyx_client_authlistener clientAuthlistener = new SrpKeyx_client_authlistener(this);
			Assert.Equal(AllJoyn.QStatus.OK, clientBus.EnablePeerSecurity("ALLJOYN_SRP_KEYX", clientAuthlistener, null, false));
			clientBus.ClearKeyStore();

			// create+activate the interface
			AllJoyn.QStatus status;
			AllJoyn.InterfaceDescription iFace = null;
			Assert.Equal(AllJoyn.QStatus.OK, clientBus.CreateInterface(INTERFACE_NAME, out iFace));
			Assert.NotNull(iFace);

			Assert.Equal(AllJoyn.QStatus.OK, iFace.AddMethod("ping", "s", "s", "in,out"));
			iFace.Activate();

			AllJoyn.ProxyBusObject proxyBusObject = new AllJoyn.ProxyBusObject(clientBus, WELLKNOWN_NAME, OBJECT_PATH, 0, true);
			Assert.NotNull(proxyBusObject);

			Assert.True(proxyBusObject.IsSecure);

			Assert.Equal(AllJoyn.QStatus.OK, proxyBusObject.AddInterface(iFace));

			AllJoyn.MsgArg input = new AllJoyn.MsgArg("s", "AllJoyn");
			AllJoyn.Message replyMsg = new AllJoyn.Message(clientBus);
			status = proxyBusObject.MethodCall(INTERFACE_NAME, "ping", input, replyMsg, 5000, 0);
			Assert.Equal(AllJoyn.QStatus.OK, status);
			Assert.Equal("AllJoyn", (string)replyMsg[0]);

			Assert.True(authflags.requestCreds_service);
			Assert.True(authflags.authComplete_serivce);

			Assert.True(authflags.requestCreds_client);
			Assert.True(authflags.authComplete_client);

			clientAuthlistener.Dispose();
			serviceAuthlistener.Dispose();

			busObject.Dispose();

			proxyBusObject.Dispose();
		}

		[Fact, Trait("objsecurity", "test")]
		public void interface_security_policy_off_secureobject()
		{
			AllJoyn.InterfaceDescription service_intf = null;
			Assert.Equal(AllJoyn.QStatus.OK, serviceBus.CreateInterface(INTERFACE_NAME, AllJoyn.InterfaceDescription.SecurityPolicy.Off, out service_intf));
			Assert.Equal(AllJoyn.QStatus.OK, service_intf.AddMethod("ping", "s", "s", "in,out"));
			service_intf.Activate();

			ResetAuthFlags();
			clientBus.ClearKeyStore();

			SrpKeyx_service_authlistener serviceAuthlistener = new SrpKeyx_service_authlistener(this);
			Assert.Equal(AllJoyn.QStatus.OK, serviceBus.EnablePeerSecurity("ALLJOYN_SRP_KEYX", serviceAuthlistener, null, false));

			serviceBus.ClearKeyStore();

			Assert.Equal(AllJoyn.QStatus.OK, SetUpAuthService());

			SrpKeyx_client_authlistener clientAuthlistener = new SrpKeyx_client_authlistener(this);
			Assert.Equal(AllJoyn.QStatus.OK, clientBus.EnablePeerSecurity("ALLJOYN_SRP_KEYX", clientAuthlistener, null, false));
			clientBus.ClearKeyStore();

			// create+activate the interface
			AllJoyn.QStatus status;
			AllJoyn.InterfaceDescription iFace = null;
			Assert.Equal(AllJoyn.QStatus.OK, clientBus.CreateInterface(INTERFACE_NAME, AllJoyn.InterfaceDescription.SecurityPolicy.Off, out iFace));
			Assert.NotNull(iFace);

			Assert.Equal(AllJoyn.QStatus.OK, iFace.AddMethod("ping", "s", "s", "in,out"));
			iFace.Activate();

			AllJoyn.ProxyBusObject proxyBusObject = new AllJoyn.ProxyBusObject(clientBus, WELLKNOWN_NAME, OBJECT_PATH, 0, true);
			Assert.NotNull(proxyBusObject);

			Assert.True(proxyBusObject.IsSecure);

			Assert.Equal(AllJoyn.QStatus.OK, proxyBusObject.AddInterface(iFace));

			AllJoyn.MsgArg input = new AllJoyn.MsgArg("s", "AllJoyn");
			AllJoyn.Message replyMsg = new AllJoyn.Message(clientBus);
			status = proxyBusObject.MethodCall(INTERFACE_NAME, "ping", input, replyMsg, 5000, 0);
			Assert.Equal(AllJoyn.QStatus.OK, status);
			Assert.Equal("AllJoyn", (string)replyMsg[0]);

			Assert.False(authflags.requestCreds_service);
			Assert.False(authflags.authComplete_serivce);

			Assert.False(authflags.requestCreds_client);
			Assert.False(authflags.authComplete_client);

			clientAuthlistener.Dispose();
			serviceAuthlistener.Dispose();

			busObject.Dispose();

			proxyBusObject.Dispose();
		}
	}
}

File: alljoyn_c/unit_test/ObjectSecurityTest.cc
/******************************************************************************
 * Copyright (c) 2013-2014, AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 ******************************************************************************/

#include <gtest/gtest.h>

#include "ajTestCommon.h"

#include <alljoyn_c/BusAttachment.h>
#include <alljoyn_c/DBusStdDefines.h>
#include <alljoyn_c/AuthListener.h>
#include <alljoyn_c/Status.h>

#include <qcc/platform.h>
#include <qcc/Thread.h>

static const char* INTERFACE_NAME = "org.alljoyn.test.c.authlistener";
static const char* OBJECT_NAME = "org.alljoyn.test.c.authlistener";
static const char* OBJECT_PATH = "/org/alljoyn/test";

static QCC_BOOL name_owner_changed_flag = QCC_FALSE;

/* NameOwnerChanged callback */
static void AJ_CALL name_owner_changed(const void* context, const char* busName, const char* previousOwner, const char* newOwner)
{
    if (strcmp(busName, OBJECT_NAME) == 0) {
        name_owner_changed_flag = QCC_TRUE;
    }
}

/* Exposed methods */
static void AJ_CALL ping_method(alljoyn_busobject bus, const alljoyn_interfacedescription_member* member, alljoyn_message msg)
{
    alljoyn_msgarg outArg = alljoyn_msgarg_create();
    alljoyn_msgarg inArg = alljoyn_message_getarg(msg, 0);
    const char* str;
    alljoyn_msgarg_get(inArg, "s", &str);
    alljoyn_msgarg_set(outArg, "s", str);
    QStatus status = alljoyn_busobject_methodreply_args(bus, msg, outArg, 1);
    EXPECT_EQ(ER_OK, status) << "  Actual Status: " << QCC_StatusText(status);
    alljoyn_msgarg_destroy(outArg);
}

static QCC_BOOL requestcredentials_service_flag = QCC_FALSE;
static QCC_BOOL authenticationcomplete_service_flag = QCC_FALSE;

static QCC_BOOL requestcredentials_client_flag = QCC_FALSE;
static QCC_BOOL authenticationcomplete_client_flag = QCC_FALSE;

class ObjectSecurityTest : public testing::Test {
  public:
    virtual void SetUp() {
        /* initilize test fixture values to known start value */
        status = ER_FAIL;
        servicebus = NULL;
        clientbus = NULL;

        testObj = NULL;
        buslistener = NULL;

        /* set up the service bus */
        servicebus = alljoyn_busattachment_create("AuthListenerTestService", false);
        status = alljoyn_busattachment_start(servicebus);
        EXPECT_EQ(ER_OK, status) << "  Actual Status: " << QCC_StatusText(status);
        status = alljoyn_busattachment_connect(servicebus, ajn::getConnectArg().c_str());
        EXPECT_EQ(ER_OK, status) << "  Actual Status: " << QCC_StatusText(status);

        clientbus = alljoyn_busattachment_create("AuthListenerTestClient", false);
        status = alljoyn_busattachment_start(clientbus);
        EXPECT_EQ(ER_OK, status) << "  Actual Status: " << QCC_StatusText(status);
        status = alljoyn_busattachment_connect(clientbus, ajn::getConnectArg().c_str());
        EXPECT_EQ(ER_OK, status) << "  Actual Status: " << QCC_StatusText(status);
    }

    virtual void TearDown() {
        alljoyn_busattachment_stop(servicebus);
        alljoyn_busattachment_join(servicebus);
        alljoyn_busattachment_stop(clientbus);
        alljoyn_busattachment_join(clientbus);
        alljoyn_busattachment_destroy(servicebus);
        alljoyn_busattachment_destroy(clientbus);
        alljoyn_buslistener_destroy(buslistener);
        alljoyn_busobject_destroy(testObj);
    }

    void SetUpAuthServiceWithSecureBusObject() {
        /* register bus listener */
        alljoyn_buslistener_callbacks buslistenerCbs = {
            NULL,
            NULL,
            NULL,
            NULL,
            &name_owner_changed,
            NULL,
            NULL,
            NULL
        };
        buslistener = alljoyn_buslistener_create(&buslistenerCbs, NULL);
        alljoyn_busattachment_registerbuslistener(servicebus, buslistener);

        /* Set up bus object */
        alljoyn_busobject_callbacks busObjCbs = {
            NULL,
            NULL,
            NULL,
            NULL
        };
        testObj = alljoyn_busobject_create(OBJECT_PATH, QCC_FALSE, &busObjCbs, NULL);
        const alljoyn_interfacedescription exampleIntf = alljoyn_busattachment_getinterface(servicebus, INTERFACE_NAME);
        ASSERT_TRUE(exampleIntf);

        status = alljoyn_busobject_addinterface(testObj, exampleIntf);
        EXPECT_EQ(ER_OK, status) << "  Actual Status: " << QCC_StatusText(status);

        /* register method handlers */
        alljoyn_interfacedescription_member ping_member;
        QCC_BOOL foundMember = alljoyn_interfacedescription_getmember(exampleIntf, "ping", &ping_member);
        EXPECT_TRUE(foundMember);

        /* add methodhandlers */
        alljoyn_busobject_methodentry methodEntries[] = {
            { &ping_member, ping_method },
        };
        status = alljoyn_busobject_addmethodhandlers(testObj, methodEntries, sizeof(methodEntries) / sizeof(methodEntries[0]));
        EXPECT_EQ(ER_OK, status) << "  Actual Status: " << QCC_StatusText(status);

        // the bus object has been registered as a secure busobject
        status = alljoyn_busattachment_registerbusobject_secure(servicebus, testObj);
        EXPECT_EQ(ER_OK, status) << "  Actual Status: " << QCC_StatusText(status);

        EXPECT_TRUE(alljoyn_busobject_issecure(testObj));
        name_owner_changed_flag = QCC_FALSE;

        /* request name */
        uint32_t flags = DBUS_NAME_FLAG_REPLACE_EXISTING | DBUS_NAME_FLAG_DO_NOT_QUEUE;
        status = alljoyn_busattachment_requestname(servicebus, OBJECT_NAME, flags);
        EXPECT_EQ(ER_OK, status) << "  Actual Status: " << QCC_StatusText(status);
        for (size_t i = 0; i < 200; ++i) {
            if (name_owner_changed_flag) {
                break;
            }
            qcc::Sleep(5);
        }
        EXPECT_TRUE(name_owner_changed_flag);
    }

    void SetupAuthClientWitSecureProxyBusObject() {
        // make a secure proxybusobject
        alljoyn_proxybusobject proxyObj = alljoyn_proxybusobject_create_secure(clientbus, OBJECT_NAME, OBJECT_PATH, 0);
        EXPECT_TRUE(proxyObj);
        EXPECT_TRUE(alljoyn_proxybusobject_issecure(proxyObj));
        status = alljoyn_proxybusobject_introspectremoteobject(proxyObj);
        EXPECT_EQ(ER_OK, status) << "  Actual Status: " << QCC_StatusText(status);

        alljoyn_message reply = alljoyn_message_create(clientbus);
        alljoyn_msgarg input = alljoyn_msgarg_create_and_set("s", "AllJoyn");
        status = alljoyn_proxybusobject_methodcall(proxyObj, INTERFACE_NAME, "ping", input, 1, reply, ALLJOYN_MESSAGE_DEFAULT_TIMEOUT, 0);
        EXPECT_EQ(ER_OK, status) << "  Actual Status: " << QCC_StatusText(status);
        const char* str;
        alljoyn_msgarg_get(alljoyn_message_getarg(reply, 0), "s", &str);
        EXPECT_STREQ("AllJoyn", str);

        alljoyn_message_destroy(reply);
        alljoyn_msgarg_destroy(input);
        alljoyn_proxybusobject_destroy(proxyObj);

    }

    void ResetAuthFlags() {
        requestcredentials_service_flag = QCC_FALSE;
        authenticationcomplete_service_flag = QCC_FALSE;

        requestcredentials_client_flag = QCC_FALSE;
        authenticationcomplete_client_flag = QCC_FALSE;
    }
    QStatus status;
    alljoyn_busattachment servicebus;
    alljoyn_busattachment clientbus;

    alljoyn_busobject testObj;
    alljoyn_buslistener buslistener;
};


/* AuthListener callback functions*/
static QCC_BOOL AJ_CALL authlistener_requestcredentials_service_srp_keyx(const void* context, const char* authMechanism,
                                                                         const char* peerName, uint16_t authCount,
                                                                         const char* userName, uint16_t credMask,
                                                                         alljoyn_credentials credentials) {
    EXPECT_STREQ("ALLJOYN_SRP_KEYX", authMechanism);
    if (credMask & ALLJOYN_CRED_PASSWORD) {
        alljoyn_credentials_setpassword(credentials, "ABCDEFGH");
    }
    requestcredentials_service_flag = QCC_TRUE;
    return QCC_TRUE;
}

static void AJ_CALL alljoyn_authlistener_authenticationcomplete_service_srp_keyx(const void* context, const char* authMechanism,
                                                                                 const char* peerName, QCC_BOOL success) {
    EXPECT_TRUE(success);
    authenticationcomplete_service_flag = QCC_TRUE;
}


static QCC_BOOL AJ_CALL authlistener_requestcredentials_client_srp_keyx(const void* context, const char* authMechanism,
                                                                        const char* peerName, uint16_t authCount,
                                                                        const char* userName, uint16_t credMask,
                                                                        alljoyn_credentials credentials) {
    EXPECT_STREQ("ALLJOYN_SRP_KEYX", authMechanism);
    if (credMask & ALLJOYN_CRED_PASSWORD) {
        alljoyn_credentials_setpassword(credentials, "ABCDEFGH");
    }
    requestcredentials_client_flag = QCC_TRUE;
    return QCC_TRUE;
}

static void AJ_CALL alljoyn_authlistener_authenticationcomplete_client_srp_keyx(const void* context, const char* authMechanism,
                                                                                const char* peerName, QCC_BOOL success) {
    EXPECT_TRUE(success);
    authenticationcomplete_client_flag = QCC_TRUE;
}


TEST_F(ObjectSecurityTest, insecure_interface_sercure_object) {
    //create an interface that is not marked with any security
    alljoyn_interfacedescription service_intf;
    status = alljoyn_busattachment_createinterface(servicebus, INTERFACE_NAME, &service_intf);
    EXPECT_EQ(ER_OK, status) << "  Actual Status: " << QCC_StatusText(status);
    ASSERT_TRUE(service_intf != NULL);
    status = alljoyn_interfacedescription_addmember(service_intf, ALLJOYN_MESSAGE_METHOD_CALL, "ping", "s", "s", "in,out", 0);
    EXPECT_EQ(ER_OK, status) << "  Actual Status: " << QCC_StatusText(status);
    alljoyn_interfacedescription_activate(service_intf);

    //make sure all the auth flags are in a known state
    ResetAuthFlags();

    /*
     * if an interface is created without security is should default to an
     * inherit security policy.
     */
    EXPECT_EQ(AJ_IFC_SECURITY_INHERIT, alljoyn_interfacedescription_getsecuritypolicy(service_intf));

    /*
     * clear the busattachment keystore to make sure authentication is performed
     * if needed.
     */
    alljoyn_busattachment_clearkeystore(clientbus);

    /* set up the service */
    alljoyn_authlistener_callbacks authlistener_cb_service = {
        authlistener_requestcredentials_service_srp_keyx,     //requestcredentials
        NULL,     //verifycredentials
        NULL,     //securityviolation
        alljoyn_authlistener_authenticationcomplete_service_srp_keyx     //authenticationcomplete
    };

    alljoyn_authlistener serviceauthlistener = alljoyn_authlistener_create(&authlistener_cb_service, NULL);

    status = alljoyn_busattachment_enablepeersecurity(servicebus, "ALLJOYN_SRP_KEYX", serviceauthlistener, NULL, QCC_FALSE);
    EXPECT_EQ(ER_OK, status) << "  Actual Status: " << QCC_StatusText(status);
    /* Clear the Keystore between runs */
    alljoyn_busattachment_clearkeystore(servicebus);

    SetUpAuthServiceWithSecureBusObject();

    /* set up the client */
    alljoyn_authlistener_callbacks authlistener_cb_client = {
        authlistener_requestcredentials_client_srp_keyx,     //requestcredentials
        NULL,     //verifycredentials
        NULL,     //securityviolation
        alljoyn_authlistener_authenticationcomplete_client_srp_keyx     //authenticationcomplete
    };

    alljoyn_authlistener clientauthlistener = alljoyn_authlistener_create(&authlistener_cb_client, NULL);

    status = alljoyn_busattachment_enablepeersecurity(clientbus, "ALLJOYN_SRP_KEYX", clientauthlistener, NULL, QCC_FALSE);
    EXPECT_EQ(ER_OK, status) << "  Actual Status: " << QCC_StatusText(status);
    /* Clear the Keystore between runs */
    alljoyn_busattachment_clearkeystore(clientbus);

    SetupAuthClientWitSecureProxyBusObject();

    EXPECT_TRUE(requestcredentials_service_flag);
    EXPECT_TRUE(authenticationcomplete_service_flag);

    EXPECT_TRUE(requestcredentials_client_flag);
    EXPECT_TRUE(authenticationcomplete_client_flag);

    alljoyn_authlistener_destroy(serviceauthlistener);
    alljoyn_authlistener_destroy(clientauthlistener);
}

TEST_F(ObjectSecurityTest, interface_security_off_sercure_object) {
    //create an interface that is not marked with any security
    alljoyn_interfacedescription service_intf;
    status = alljoyn_busattachment_createinterface_secure(servicebus, INTERFACE_NAME, &service_intf, AJ_IFC_SECURITY_OFF);
    EXPECT_EQ(ER_OK, status) << "  Actual Status: " << QCC_StatusText(status);
    ASSERT_TRUE(service_intf != NULL);
    status = alljoyn_interfacedescription_addmember(service_intf, ALLJOYN_MESSAGE_METHOD_CALL, "ping", "s", "s", "in,out", 0);
    EXPECT_EQ(ER_OK, status) << "  Actual Status: " << QCC_StatusText(status);
    alljoyn_interfacedescription_activate(service_intf);

    //make sure all the auth flags are in a known state
    ResetAuthFlags();

    /*
     * if an interface is created with off security policy.
     */
    EXPECT_EQ(AJ_IFC_SECURITY_OFF, alljoyn_interfacedescription_getsecuritypolicy(service_intf));

    /*
     * clear the busattachment keystore to make sure authentication is performed
     * if needed.
     */
    alljoyn_busattachment_clearkeystore(clientbus);

    /* set up the service */
    alljoyn_authlistener_callbacks authlistener_cb_service = {
        authlistener_requestcredentials_service_srp_keyx,     //requestcredentials
        NULL,     //verifycredentials
        NULL,     //securityviolation
        alljoyn_authlistener_authenticationcomplete_service_srp_keyx     //authenticationcomplete
    };

    alljoyn_authlistener serviceauthlistener = alljoyn_authlistener_create(&authlistener_cb_service, NULL);

    status = alljoyn_busattachment_enablepeersecurity(servicebus, "ALLJOYN_SRP_KEYX", serviceauthlistener, NULL, QCC_FALSE);
    EXPECT_EQ(ER_OK, status) << "  Actual Status: " << QCC_StatusText(status);
    /* Clear the Keystore between runs */
    alljoyn_busattachment_clearkeystore(servicebus);

    SetUpAuthServiceWithSecureBusObject();

    /* set up the client */
    alljoyn_authlistener_callbacks authlistener_cb_client = {
        authlistener_requestcredentials_client_srp_keyx,     //requestcredentials
        NULL,     //verifycredentials
        NULL,     //securityviolation
        alljoyn_authlistener_authenticationcomplete_client_srp_keyx     //authenticationcomplete
    };

    alljoyn_authlistener clientauthlistener = alljoyn_authlistener_create(&authlistener_cb_client, NULL);

    status = alljoyn_busattachment_enablepeersecurity(clientbus, "ALLJOYN_SRP_KEYX", clientauthlistener, NULL, QCC_FALSE);
    EXPECT_EQ(ER_OK, status) << "  Actual Status: " << QCC_StatusText(status);
    /* Clear the Keystore between runs */
    alljoyn_busattachment_clearkeystore(clientbus);

    SetupAuthClientWitSecureProxyBusObject();

    EXPECT_FALSE(requestcredentials_service_flag);
    EXPECT_FALSE(authenticationcomplete_service_flag);

    EXPECT_FALSE(requestcredentials_client_flag);
    EXPECT_FALSE(authenticationcomplete_client_flag);

    alljoyn_authlistener_destroy(serviceauthlistener);
    alljoyn_authlistener_destroy(clientauthlistener);
}
